File: backend/app/utils/redaction.py
"""Shared secret redaction.

A single best-effort scrubber for secret shapes that could appear verbatim in text
we did not author — library log lines (asyncssh logs every command it runs at INFO,
so a command embedding a credential inline would otherwise leak) and Sentry event
payloads. Both the logging sink and the monitoring `before_send` hook call
`redact_secrets` so there is one source of truth for what a secret looks like.

These are a safety net, not a guarantee: they only catch the shapes listed here. Our
own code avoids logging secrets in the first place.
"""

import re

# GitHub tokens (ghp_/gho_/ghs_/ghr_/ghu_ ...).
_GITHUB_TOKEN = re.compile(r"\bgh[posru]_[A-Za-z0-9]{20,}\b")
_BEARER = re.compile(r"(?i)(authorization:\s*bearer\s+)\S+")
# The password half of a `user:password` credential fed to chpasswd. Passwords never
# contain whitespace, quotes, or a pipe, so the match stops cleanly at the shell
# delimiter; the length floor avoids masking unix `user:group` pairs (chown root:root).
_CHPASSWD_CRED = re.compile(r"(\b[\w.-]+:)([^\s'\"|]{8,})")


def redact_secrets(message: str) -> str:
    """Mask known secret shapes in a message, preserving surrounding context so the
    line still shows what ran. Best-effort: only the patterns above are covered."""
    if "chpasswd" in message:
        message = _CHPASSWD_CRED.sub(r"\1***", message)
    message = _GITHUB_TOKEN.sub("***", message)
    message = _BEARER.sub(r"\1***", message)
    return message

File: frontend/src/lib/redaction.ts
// Shared secret redaction (frontend mirror of backend app/utils/redaction.py).
//
// A single best-effort scrubber for secret shapes that could appear verbatim in
// text we did not author — an error message or exception value that happens to
// carry a token or inline credential. The Sentry `beforeSend` hook calls
// `redactSecrets` so there is one source of truth for what a secret looks like,
// the same way the backend log sink and Sentry `before_send` both call it.
//
// These are a safety net, not a guarantee: they only catch the shapes listed
// here. Our own code avoids putting secrets into messages in the first place.

// GitHub tokens (ghp_/gho_/ghs_/ghr_/ghu_ ...).
const GITHUB_TOKEN = /\bgh[posru]_[A-Za-z0-9]{20,}\b/g
const BEARER = /(authorization:\s*bearer\s+)\S+/gi
// The password half of a `user:password` credential fed to chpasswd. Passwords
// never contain whitespace, quotes, or a pipe, so the match stops cleanly at the
// shell delimiter; the length floor avoids masking unix `user:group` pairs.
const CHPASSWD_CRED = /(\b[\w.-]+:)([^\s'"|]{8,})/g

// Mask known secret shapes in a message, preserving surrounding context so the
// line still shows what happened. Best-effort: only the patterns above are covered.
export function redactSecrets(message: string): string {
  let out = message
  if (out.includes('chpasswd')) {
    out = out.replace(CHPASSWD_CRED, '$1***')
  }
  out = out.replace(GITHUB_TOKEN, '***')
  out = out.replace(BEARER, '$1***')
  return out
}
